### ISSUE TITLE
exchange() failing with The request has an invalid grant parameter: code

### ISSUE THREAD

**Issue by ericblade** [amz-tools/amazon-sp-api#182]
I'm not sure if this is something that came up with 0.8.0 release, or if it's amazon side, or if something else entirely is wrong, but.. this started recently...  when I do an exchange(), i always get invalid grant parameter: code

````
async function getSpApiForOAuth(spapiOauthCode) {
    if (!spapiOauthCode) {
        throw new Error('Please re-authorize with Amazon in Settings');
    }
    // given the oauthcode, create a grantless spapi instance to exchange for a reauth token
    const spapi = new SellingPartnerAPI({
        region: 'na',
        options: {
            only_grantless_operations: true,
        },
        credentials: {
            SELLING_PARTNER_APP_CLIENT_ID,
            SELLING_PARTNER_APP_CLIENT_SECRET,
            AWS_SECRET_ACCESS_KEY,
            AWS_ACCESS_KEY_ID,
            AWS_SELLING_PARTNER_ROLE,
        },
    });
    // exchange for re-auth
    const res = await (async function() {
        try {
            const x = await spapi.exchange(spapiOauthCode);
            console.warn('* spapi exchange returns', x);
            return x;
        } catch (e) {
            console.error('* oauth exchange failed', e);
            if (e.code === 'invalid_grant') {
                // throw an error the client can recognize uniquely that auth has failed/expired
                throw new Error('Please re-authorize with Amazon in Settings');
            }
            throw e;
        }
    }());
    // create a new spapi interface with the refresh token and hang onto it for future requests
    const actual = new SellingPartnerAPI({
        region: 'na',
        refresh_token: res.refresh_token,
        credentials: {
            SELLING_PARTNER_APP_CLIENT_ID,
            SELLING_PARTNER_APP_CLIENT_SECRET,
            AWS_SECRET_ACCESS_KEY,
            AWS_ACCESS_KEY_ID,
            AWS_SELLING_PARTNER_ROLE,
        },
    });
    spapiCache.set(res.refresh_token, actual);
    return actual;
}
````

am i doing something wrong here?

**Comment by amz-tools** [comment]
@ericblade One guess: Is you authorized app a hybrid SP/MWS app? As MWS is shut down now maybe this might be the issue?

**Comment by ericblade** [comment]
Yes, it is a hybrid, but MWS doesn't shutdown until April 2024 at the least.  
Not sure what specifically that would have to do with it, either.

It was working until mid December at least, which is when I started doing some work on the app again (coincidentally, updating some things to use SPAPI), so I thought perhaps pulling in new versions of this lib might've caused it, but going backwards to priors didn't seem to get there.

I don't really have any ideas as to what might be going on.  Or, for that matter, how to diagnose it any further :-S

**Comment by nicolasheady** [comment]
@ericblade I have the same issue...did you solve it?

**Comment by ericblade** [comment]
I do not have any further information at t his time @nicolasheady .. this is not my day job, so I haven't really had much time to poke at it.  It is something that i need to get together, though, I'd like to be able to get this fixed so I can go public someday :D 

If you happen to have any ideas/find anything out, I'd love to hear.  I will also come back here and report back if I find anything that seems to get through it.

**Comment by ankush981** [comment]
Came here to say that this was happening to me as well. However, it happened when I was copying the query string from my production domain to localhost for local development and testing. As soon as decided to test on the production server, the problem went away! Hard to say if this indeed was the cause, and I'm to lazy to modify my local DNS to point to a non-"localhost" domain and then pass that as the redirect uri, so I'm happy for now! 😄

**Comment by ericblade** [comment]
Coming back to this almost a year later.  :-D  This is now failing even all the way back to the previously working 0.6.6 version.  

I haven't the slightest idea.  It was working all the way through till now, when my Amazon credentials needed updating, and now it's just all failing.

**Comment by ericblade** [comment]
We should be feeding it the Amazon Developer security profile settings, yes?

The Client ID and Client Secret from this page? 
![image](https://github.com/user-attachments/assets/7e8419cf-c816-484a-930e-3edf8f08df6b)

(clipped before the id and secret)

**Comment by amz-tools** [comment]
@ericblade Nowadays the client only needs the `SELLING_PARTNER_APP_CLIENT_ID` and `SELLING_PARTNER_APP_CLIENT_SECRET` credentials. `AWS_SECRET_ACCESS_KEY`, `AWS_ACCESS_KEY_ID` and `AWS_SELLING_PARTNER_ROLE ` are NOT needed anymore.

ID and Secret can be found either in the new solution provider portal if you have already switched to there, otherwise you can find it in your Sellercentral when you go to "Apps and Services" --> "Develop Apps" in the "LWA credentials" section, see screenshot.

No idea though if this helps you or if you have already set it up correct. :-) 

<img width="672" alt="Bildschirmfoto 2024-11-28 um 09 55 19" src="https://github.com/user-attachments/assets/9551521a-53ff-4ff3-b010-83fc64a7d23a">

**Comment by ericblade** [comment]
grrr.  Amazon does some really weird things.  When I go to the LWA page, it forwards me to Amazon Developer, which gives me a different pair of credentials, than if I go to Apps and Services, *then* LWA.  

Either way, neither credential gives me anything but "invalid_grant".  Removing the AWS_* doesn't seem to make any change either.  This "security through obscure error messages" is not helpful, Amazon.

**Comment by ericblade** [comment]
Noting that my original error was " invalid grant parameter : code" and now it's "invalid_grant", guessing that's probably an amazon change
